clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/OperationsService.java

```diff
@@ -0,0 +1,252 @@
+import com.google.common.base.Splitter;
+import com.netflix.spectator.api.Id;
+import com.netflix.spectator.api.Registry;
+import com.netflix.spinnaker.clouddriver.deploy.DescriptionAuthorizer;
+import com.netflix.spinnaker.clouddriver.deploy.DescriptionValidationErrors;
+import com.netflix.spinnaker.clouddriver.deploy.DescriptionValidationException;
+import com.netflix.spinnaker.clouddriver.deploy.DescriptionValidator;
+import com.netflix.spinnaker.clouddriver.security.AccountCredentials;
+import com.netflix.spinnaker.clouddriver.security.AccountCredentialsRepository;
+import com.netflix.spinnaker.clouddriver.security.AllowedAccountsValidator;
+import com.netflix.spinnaker.clouddriver.security.ProviderVersion;
+import com.netflix.spinnaker.kork.exceptions.SystemException;
+import com.netflix.spinnaker.security.AuthenticatedRequest;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.Data;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.core.ResolvableType;
+import org.springframework.validation.Errors;
+
+@Slf4j
+public class OperationsService {
+
+  private final Splitter COMMA_SPLITTER = Splitter.on(",");
+
+  private final AtomicOperationsRegistry atomicOperationsRegistry;
+  private final DescriptionAuthorizer descriptionAuthorizer;
+  private final Collection<AllowedAccountsValidator> allowedAccountValidators;
+  private final List<AtomicOperationDescriptionPreProcessor>
+      atomicOperationDescriptionPreProcessors;
+  private final AccountCredentialsRepository accountCredentialsRepository;
+  private final Registry registry;
+  private final ObjectMapper objectMapper;
+
+  private final Id validationErrorsCounterId;
+
+  public OperationsService(
+      AtomicOperationsRegistry atomicOperationsRegistry,
+      DescriptionAuthorizer descriptionAuthorizer,
+      Optional<Collection<AllowedAccountsValidator>> allowedAccountValidators,
+      Optional<List<AtomicOperationDescriptionPreProcessor>>
+          atomicOperationDescriptionPreProcessors,
+      AccountCredentialsRepository accountCredentialsRepository,
+      Registry registry,
+      ObjectMapper objectMapper) {
+    this.atomicOperationsRegistry = atomicOperationsRegistry;
+    this.descriptionAuthorizer = descriptionAuthorizer;
+    this.allowedAccountValidators = allowedAccountValidators.orElse(Collections.emptyList());
+    this.atomicOperationDescriptionPreProcessors =
+        atomicOperationDescriptionPreProcessors.orElse(Collections.emptyList());
+    this.accountCredentialsRepository = accountCredentialsRepository;
+    this.registry = registry;
+    this.objectMapper = objectMapper;
+
+    validationErrorsCounterId = registry.createId("validationErrors");
```

Comment:
the error counter name could be a little more descriptive? `operationValidationErrors`?

Reply:
This isn't new code, I just moved it out from the controller, so I don't want to change the metric names, even if they're pretty rubbish. :(

We probably should have a plan around how to change metric names, though.

Reply:
oh :/